include/rspq.h

```diff
@@ -388,6 +401,109 @@ void* rspq_overlay_get_state(rsp_ucode_t *overlay_ucode);
+    volatile uint32_t *pointer;           ///< current pointer into the RSP queue
+    volatile uint32_t *first;             ///< pointer to the first word of the command
+    bool is_first;                        ///< true if we are waiting for the first argument word
+} rspq_write_t;
+
+/**
+ * @brief Begin writing a new command into the RSP queue.
+ * 
+ * This command initiates a sequence to enqueue a new command into the RSP
+ * queue. Call this command passing the overlay ID and command ID of the command
+ * to create. Then, call #rspq_write_arg once per each argument word that
+ * composes the command. Finally, call #rspq_write_end to finalize and enqueue
+ * the command.
+ * 
+ * A sequence made by #rspq_write_begin, #rspq_write_arg, #rspq_write_end is
+ * functionally equivalent to a call to #rspq_write, but it allows to
+ * create bigger commands, and might better fit some situations where arguments
+ * are calculated on the fly. Performance-wise, the code generated by
+ * #rspq_write_begin + #rspq_write_arg + #rspq_write_end should be very similar
+ * to a single call to #rspq_write, though just a bit slower. It is advisable
+ * to use #rspq_write whenever possible.
+ * 
+ * Make sure to read the documentation of #rspq_write as well for further
+ * details.
+ * 
+ * @param      ovl_id    The overlay ID of the command to enqueue. Notice that
+ *                       this must be a value preshifted by 28, as returned
+ *                       by #rspq_overlay_register.
+ * @param      cmd_id    Index of the command to call, within the overlay.
+ * @param      size      The size of the commands in 32-bit words
+ * @returns              A write cursor, that must be passed to #rspq_write_arg
+ *                       and #rspq_write_end
+ * 
+ * @see #rspq_write_arg
+ * @see #rspq_write_end
+ * @see #rspq_write
+ */
+inline rspq_write_t rspq_write_begin(uint32_t ovl_id, uint32_t cmd_id, int size) {
+    extern volatile uint32_t *rspq_cur_pointer, *rspq_cur_sentinel;
+    extern void rspq_next_buffer(void);
+
+    if (__builtin_expect(rspq_cur_pointer > rspq_cur_sentinel - size, 0))
+        rspq_next_buffer();
+
+    volatile uint32_t *cur = rspq_cur_pointer;
+    rspq_cur_pointer += size;
+
+    return (rspq_write_t){
+        .first_word = ovl_id + (cmd_id<<24),
+        .pointer = cur + 1,
+        .first = cur,
+        .is_first = 1
+    };
+}
+
+/**
+ * @brief Add one argument to the command being enqueued.
+ * 
+ * This function adds one more argument to the command currently being
+ * enqueued. This function must be called after #rspq_write_begin; it should
+ * be called multiple times (one per argument word), and then #rspq_write_end
+ * should be called to terminate enqueuing the command.
+ * 
+ * See also #rspq_write for a more straightforward API for command enqueuing.
+ * 
+ * @param       w       The write cursor (returned by #rspq_write_begin)
+ * @param       value   New 32-bit argument word to add to the command.
```

Comment:
> It might be nice to include the first argument restriction we have here as well. I mean the first byte should not be used for the first argument.

Sure, added.

> What about including the first argument on rspq_write_begin instead, to be more consistent with rspq_write API? We could get rid of the additional is_first state as well. This is not something I'm against but just thinking out loud.

I think it would make the API less orthogonal: for instance, if you check rdp_triangle in the current gfx-rdp branch (https://github.com/rasky/libdragon/blob/4ff51e9a487a0885a375893827c3e44b510eae56/src/rdpq/rdpq.c#L669) you will see that the first argument is generated within the `__rdpq_write_edge_coeffs` function, so forcing the first argument in rspq_write_begin would make it for a more awkward implementation.

> Also do you think we should assert if the first byte of the first arg is not zero?

If we do it, we should do it also in `rspq_write`. I think I've had a few cases where I actually abused the system already because the assert was not there, but I'm not strongly opposed to adding it. It would have a measurable code-size / performance impact though (unless we limit it to compile-time constants). I think we could discuss it as part of a separate PR anyway.